presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ReplaceConstantVariableReferencesWithConstants.java

```diff
@@ -0,0 +1,431 @@
+    {
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionAndTypeManager is null");
+    }
+
+    private static boolean isSupportedType(RowExpression rowExpression)
+    {
+        return SUPPORTED_JOIN_KEY_TYPE.contains(rowExpression.getType()) || rowExpression.getType() instanceof VarcharType;
+    }
+
+    @Override
+    public PlanOptimizerResult optimize(PlanNode plan, Session session, TypeProvider types, VariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        if (isRewriteExpressionWithConstantEnabled(session)) {
+            Rewriter rewriter = new Rewriter(idAllocator, functionAndTypeManager);
+            PlanNode rewrittenPlan = plan.accept(rewriter, null).getPlanNode();
+            return PlanOptimizerResult.optimizerResult(rewrittenPlan, rewriter.isPlanChanged());
+        }
+        return PlanOptimizerResult.optimizerResult(plan, false);
+    }
+
+    private static class Rewriter
+            extends InternalPlanVisitor<PlanNodeWithConstant, Void>
+    {
+        private final PlanNodeIdAllocator idAllocator;
+        private final FunctionResolution functionResolution;
+        private boolean planChanged;
+
+        public Rewriter(PlanNodeIdAllocator idAllocator, FunctionAndTypeManager functionAndTypeManager)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.functionResolution = new FunctionResolution(functionAndTypeManager.getFunctionAndTypeResolver());
+        }
+
+        public boolean isPlanChanged()
+        {
+            return planChanged;
+        }
+
+        private PlanNodeWithConstant accept(PlanNode node)
+        {
+            return node.accept(this, null);
+        }
+
+        private PlanNodeWithConstant planAndReplace(PlanNode node, boolean keepConstantConstraint)
+        {
+            List<PlanNodeWithConstant> children = node.getSources().stream().map(this::accept).collect(toImmutableList());
+            ImmutableList.Builder<PlanNode> newSources = ImmutableList.builder();
+
+            for (PlanNodeWithConstant planNodeWithConstant : children) {
+                PlanNode child = planNodeWithConstant.getPlanNode();
+                Map<VariableReferenceExpression, ConstantExpression> constantExpressionMap = planNodeWithConstant.getConstantExpressionMap();
+                if (child.getOutputVariables().stream().noneMatch(x -> constantExpressionMap.containsKey(x))) {
+                    newSources.add(child);
+                    continue;
+                }
+                Assignments.Builder assignmentsBuilder = Assignments.builder();
+                assignmentsBuilder.putAll(child.getOutputVariables().stream().collect(toImmutableMap(identity(), x -> constantExpressionMap.containsKey(x) ? constantExpressionMap.get(x) : x)));
+                newSources.add(new ProjectNode(idAllocator.getNextId(), child, assignmentsBuilder.build()));
```

Comment:
I see that we are adding a new ProjectNode even if there was no actual plan change from the constant replacement on the child nodes. Eg : 
```
2023-11-06T14:22:32.146+0530	INFO	Query-20231106_085231_00000_ap56t-1168	com.facebook.presto.sql.planner.optimizations.LogPlanTreeOptimizer	------ --576  : Before constant folding
2023-11-06T14:22:32.146+0530	INFO	Query-20231106_085231_00000_ap56t-1168	com.facebook.presto.sql.planner.optimizations.LogPlanTreeOptimizer	Plan :
project, Assignments ({name=Slice{base=[B@70da926b, address=16, length=24}}), Outputs ([name]):
    join (INNER), Equi-join condition([nationkey = nationkey_1]), Filter (Optional.empty), Outputs([]):
        project, Assignments ({nationkey=nationkey}), Outputs ([nationkey]):
            filter (predicate = EQUAL(name, Slice{base=[B@430f0bf7, address=16, length=24}))
                scan (nation:sf1.0) Constraints [Current (Optional[{}]), Enforced(Optional[{}])]
        scan (supplier:sf1.0) Constraints [Current (Optional[{}]), Enforced(Optional[{}])]

2023-11-06T14:22:32.146+0530	INFO	Query-20231106_085231_00000_ap56t-1168	com.facebook.presto.sql.planner.optimizations.LogPlanTreeOptimizer	------ --576  : After constant folding
2023-11-06T14:22:32.146+0530	INFO	Query-20231106_085231_00000_ap56t-1168	com.facebook.presto.sql.planner.optimizations.LogPlanTreeOptimizer	Plan :
project, Assignments ({name=Slice{base=[B@70da926b, address=16, length=24}}), Outputs ([name]):
    project, Assignments ({name=Slice{base=[B@70da926b, address=16, length=24}}), Outputs ([name]):
        join (INNER), Equi-join condition([nationkey = nationkey_1]), Filter (Optional.empty), Outputs([]):
            project, Assignments ({nationkey=nationkey}), Outputs ([nationkey]):
                filter (predicate = EQUAL(name, Slice{base=[B@430f0bf7, address=16, length=24}))
                    scan (nation:sf1.0) Constraints [Current (Optional[{}]), Enforced(Optional[{}])]
            scan (supplier:sf1.0) Constraints [Current (Optional[{}]), Enforced(Optional[{}])]
```

Here we added a (duplicate extra) `project, Assignments ({name=Slice{base=[B@70da926b, address=16, length=24}}), Outputs ([name]):` node.  

I think we may be able to tweak the test for when a child node plan has really changed to avoid adding the extra Project ?

Reply:
This is an intentional design.
Based on the discussion in the threads, we decided to do only constant pull up in this optimizer, and leave the optimization to happen in other optimizers. For example, for query `select orderkey, partkey, count(*) from lineitem where orderkey=2 group by orderkey, partkey`, it will produce a plan like
aggregation <- project(orderkey=2) <- scan
then `PullConstantsAboveGroupBy` will do the optimization, where it checks for pattern above.
Hence, even if the child node does not change, we still need to add the project node, since its parent node may benefit from this optimization.
It's also the reason why I run inine projection optimizer after this optimizer, so as to simplify the plan.

Reply:
Hmm.. I'm trying to repro this behavior (of adding a duplicate Project) in a test but am unable to do so.  

I understand your point; mine was that if 1/ the current node is a ProjectNode 2/we are wrapping up it's source in another ProjectNode, we can ideally merge the assignments and skip adding the extra Project.  
I know that we end up merging/removing such duplicate Project's but it's best to avoid adding them in the first place

Reply:
This rewriter override `visitProject` method, hence when the current node is a project node, it will go through that function, and will not produce redundant project node.

Reply:
Found the root cause - When the `PlanNode node` is an `OutputNode` we can skip running `planAndReplace` on this output node and just call accept on it's source, since the OutputNode will never have a parent. Once I added : 
```
@Override
        public PlanNodeWithConstant visitOutput(OutputNode node, Void context)
        {
            accept(node.getSource());
            return new PlanNodeWithConstant(node, ImmutableMap.of());
        }
```
the duplicate Projects added went away.
We can do this for other terminal nodes as well - `TableWriter`, `TableFinish` ?

Reply:
The project with constant assignments under the output node is intentional, so that the unused variables can be pruned as early as possible.

Reply:
Maybe I'm missing something here - can you explain with an example ?

Reply:
A simple example is a query with filter, for example `select orderkey from lineitem where orderkey=3`. The plan will be like Output <- Filter -< Scan. By adding a project with constant assignment, it will be Output <- Project <- Filter <- Scan, the Project node here will create a run length encode block, but without this project, it will just be a query which does not take advantage of the constant folding here.

Reply:
So the FilterNode will have the `orderkey` as a non RLE block (given to it by the Scan). The filter is applied and the block is passed on to the Project node that we added, and a new RLE-Block-orderkey is created just to plumb it to the Output ?
Won't this be worse off memory-wise & CPU-wise than passing the original `orderkey' column as is ?  

If the parent node was a non-OutputNode that was consuming this column in some way (aggregating it etc) then adding the Project makes sense, but if its just an OutputNode, I don't see how it's better.

Maybe a microbenchmark will make this clearer and show me what I'm missing..